packages/sol/src/availabilitySlots.ts

```diff
@@ -0,0 +1,144 @@
+import { IAvailabilitySlot } from "@litespace/types";
+import { dayjs } from "@/dayjs";
+import { clone, flatten, min } from "lodash";
+
+/**
+  * Divide list of slots into sub-slots. It is used whenever a user wants to book a lesson
+  * e.g., slot from 8am-10pm, split by 30min => 8-8:30, 8:30-9, 9-9:30, 9:30-10.
+  * @param slots - list of the slots to be divided.
+  * @param duration - the duration of each subslot in minutes.
+  */
+export function getSubslotsBatch(
+  slots: IAvailabilitySlot.Slot[], 
+  duration: number,
+): IAvailabilitySlot.SubSlot[] {
+  return flatten(slots.map(slot => getSubslots(slot, duration)));
+}
+
+export function getSubslots(
+  slot: IAvailabilitySlot.Slot,
+  duration: number,
+): IAvailabilitySlot.SubSlot[] {
+  let start = dayjs.utc(slot.start);
+  let end = start.add(duration, "minutes");
+  const subslots: IAvailabilitySlot.SubSlot[] = [];
+  while (end.isBefore(dayjs.utc(slot.end)) || end.isSame(dayjs.utc(slot.end))) {
+    subslots.push({
+      parent: slot.id,
+      start: start.toISOString(),
+      end: end.toISOString(),
+    });
+    start = end;
+    end = start.add(duration, "minutes");
+  }
+  return subslots;
+}
+
+/** 
+  * checks if the passed book info can be booked in the given slot.
+  * @param slot - the slot to be partailly or fully booked.
+  * @param bookedSubslots - subslots that are already booked (to be subtracted).
+  * @param bookInfo - the time that's needed to be booked.
+  */
+export function canBook({ 
+  slot,
+  bookedSubslots,
+  bookInfo,
+}: {
+   slot: IAvailabilitySlot.Slot;
+   bookedSubslots: IAvailabilitySlot.SubSlot[];
+   bookInfo: {
+     start: string;
+     duration: number; // in minutes
+   }
+}): boolean {
+  const mainSubslots = getSubslots(slot, bookInfo.duration);
+  const filtered = mainSubslots.filter(sub => !isIntersecting(sub, bookedSubslots));
+  
+  // Check if the booking slot is a subslot of at least one of the filtered list;
+  // if so return true, otherwise return false
+  const bookingSlot: IAvailabilitySlot.SubSlot = {
+    parent: -1,
+    start: bookInfo.start,
+    end: dayjs(bookInfo.start).add(bookInfo.duration, "minutes").toISOString(),
+  };
+  for (const slot of filtered) {
+    if (isSuperSlot(slot, bookingSlot))
+      return true;
+  }
+  return false;
```

Comment:
This is not how you check if you the new slot will fit in the remaining space in the parent slot. 

Here is the steps to do this. 
1. Subtract the parent slots from all subslots. The result will be all free sub-slots. 
2. Check if the new slot (`bookInfo`) fit in one or more sub-slot. It if does fit, this mean you can book this new slot.